Markers docs

aeshky opened the pull request:
**Proposed changes**:
- Closes https://github.com/RasaHQ/rasa/issues/9748

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

usc-m commented on docs/docs/marker.mdx:
```diff
@@ -0,0 +1,280 @@
+# Markers docs
+
+# Markers
```
Do we need to mention it's an experimental feature anywhere?

aeshky replied:
Now added.

usc-m commented on docs/docs/marker.mdx:
```diff
@@ -0,0 +1,280 @@
+
+Each `operator` is a dictionary mapping an operator tag (e.g., `and`) to a list of sub markers, which can be conditions or a nested operator-conditions definition (e.g., see `marker_mood_expressed_and_name_provided`).
+
+<aside>
```
This seems like it shouldn't be here for markdown but I'm not sure if it's something that's picked up by our docs renderer

aeshky replied:
replaced it with 
```
:::note
text goes here 
:::
```
which I found in other docs.

usc-m commented on docs/docs/marker.mdx:
```diff
@@ -0,0 +1,280 @@
+
+We output the extracted information in a `.csv` file that contains the following columns:
+
+- sender_id
```
Might be an idea to format these column names somehow (code or italics?)

usc-m commented on docs/docs/marker.mdx:
```diff
@@ -0,0 +1,280 @@
+                        statistics. More precisely, the file prefix must consist of a common path
+                        plus a common file prefix, to which suffixes `-overall.csv` and `-per-
+                        session.csv` will be added automatically. (default: stats)
+  --endpoints ENDPOINTS
```
I think it's probably worth mentioning these other configuration files (endpoints config, and domain) that the tool is looking for further up in the body of the text as well - would be good for users to know about the domain validation behaviour as well as the fact it needs the endpoints config file to actually connect to a tracker store and won't work without one (and it will assume default paths if none are specified)

usc-m commented on docs/docs/marker.mdx:
```diff
@@ -0,0 +1,280 @@
+    1. overall statistics including the arithmetic mean, median, minimum, and maximum number of user turns preceding the event where the respective marker applied in any session
+    2. the number of sessions and the percentage of sessions where each marker applied at least once
+
+The results are stored in tabular format with the suffixes in the files `stats-overall.csv` and `stats-per-session.csv`. You can change the prefix `stats` via the command line.
```
It might be good to give an example of what this means, just to make it clearer

usc-m commented on docs/docs/marker.mdx:
````diff
@@ -0,0 +1,280 @@
+2cf4106e76a041dfb2d4f75bd6ae804b,0,marker_mood_expressed,7,1
+```
+
+extracted_markers.csv
````
Might be good to also give an example command here that shows how you'd produce such a file

usc-m reviewed (commented):
Looking good so far! Have some suggestions on some possible improvements

usc-m commented on docs/docs/marker.mdx:
````diff
@@ -0,0 +1,280 @@
+
+Markers are extracted from the dialogues already stored in your [tracker store](https://rasa.com/docs/rasa/next/tracker-stores). Once you've created your Marker definitions in the Makers configuration file, you can extract them by running the follow command:
+
+```jsx
````
```suggestion
```
```

usc-m replied:
Annoyingly it's hard for me to make a suggestion for you to auto-commit this because the syntax for the suggestion and the syntax for this are the same and it just suggests deleting the line 😅

aeshky replied:
I replaced it with bash. 

I was sloppy: I exported the notion page and committed the output without checking that everything made sense.

tttthomasssss commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+:::
+
+Markers allow you to describe points in a dialogue that you are interested in identifying. In Rasa, a dialogue is represented as a sequence of events, which include bot actions that were executed (`ActionExecuted`), intents that were detected (`UserUttered`), and slots that were set (`SlotSet`). A point of interest in a dialogue can thus be expressed as a logical expression that can be evaluated  ...
```
This is a fairly low-level technical description - would it make sense to have something like a 3-column figure, where in the first column we show a bunch of turns in a carbon-bot conversation, in the middle column we show the technical view on how this is translated to Rasa's event implementation and in the 3rd column we show how the success-markers steedmark a particular point in the dialogue on the basis of a yaml config and the current tracker state.

aeshky replied:
Thomas the figure is a good idea but requires some thought. I'll reserve it for the blog post 🙂

tttthomasssss commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+Markers allow you to describe points in a dialogue that you are interested in identifying. In Rasa, a dialogue is represented as a sequence of events, which include bot actions that were executed (`ActionExecuted`), intents that were detected (`UserUttered`), and slots that were set (`SlotSet`). A point of interest in a dialogue can thus be expressed as a logical expression that can be evaluated  ...
+
+There are several applications for Markers. For example, they can be used to define your bot's Key Performance Indicators (KPIs), such as dialogue completion or task success. For [Carbon Bot](https://rasa.com/blog/using-conversation-tags-to-measure-carbon-bots-success-rate/) (which helps users offset their carbon emissions from flying) dialogue completion can be defined as "all mandatory slots ha ...
```
One thing that has come up in discussions with customers again and again is that they like to measure containment (by which they usually mean dialogue completion) and track human-handoff. It might make sense to add an example around human-handoff in the docs as this is something that will a problem that is immediately familiar to them when they read the docs.

koaning replied:
> "Marking these events allows us to evaluate Carbon Bot's performance by quantifying how often it succeeds and how often it fails."

Since the word "performance" can mean many things and may be confused with machine learning stats. Does it make sense to this one to: 

> "Marking these events allows us to track how often conversations in Carbon Bot cause the effect it's designed for."

tttthomasssss commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+    and:
+        - slot_was_set: travel_departure
+        - slot_was_set: travel_destination
+        - not:
+            slot_was_set: travel_flight_class
```
Which one is the preferred way now, like here, or like in line 66 below with the `slot_was_not_set` key?

aeshky replied:
They have the same behaviour. The shorthand `slot_was_not_set`is mostly for convenience. We also have tags like `not_intent` in rules and stories, so users might attempt to try them anyway 😄

tttthomasssss commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+To describe the events that we're interested in, we provide `conditions` and `operators`.
+
+Conditions are simple expressions that describe an event, and are used like to variables in a logical expression, for example `action: utter_greet`. Operators allow you to combine conditions or sub-marker definitions which include nested operators and conditions.
```
I think `event` is ambiguous here because it was used to refer to the technical Rasa notion of "event" (`ActionExecuted`, `UserUttered`, etc) at the top, but is referred in a "something that happened at some point in the dialogue" sense here.

hsm207 replied:
@aeshky 

> and are used like to variables

You meant "and are used like variable", right?

> for example `action: utter_greet`

I don't understand this example. Is `action` the condition and `utter_greet` the event?

tttthomasssss commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,291 @@
+To define one or more marker, you need to specify a `yaml` configuration file. Here is an example:
+
+```yaml
+marker_name_provided:
+    slot_was_set: name
+marker_mood_expressed:
+    or:
+        - intent: mood_unhappy
+        - intent: mood_great
+marker_cheer_up:
+    seq:
+        - intent: mood_unhappy
+        - action: utter_cheer_up
+marker_bot_challenged:
+    at_least_once:
+        - intent: bot_challenge
+marker_mood_expressed_and_name_provided:
+    and:
+        - slot_was_set: name
+        - or:
+            - intent: mood_unhappy
+            - intent: mood_great
+```
````
For consistency reasons, would it make sense to continue using examples from carbon-bot rather than moodbot?

koaning replied:
I might prefer an example that is easy to copy/paste by the average user. That suggests that moodbot would be a nice example even though it doesn't have slots.

koaning replied:
Then again ... I suppose there's also something to be said to also add an example of an "obvious use-case". Maybe a path for "cheering up worked" vs. "cheering up failed"?

tttthomasssss commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,291 @@
+            - intent: mood_great
+```
+
+The top level contains unique user-specified names for each marker. This is a dictionary mapping marker names to marker definitions.
````
> This is a dictionary mapping marker names to marker definitions.
Do we want to expose implementation details here? I don't think this is important for the end user.

ka-bu replied:
For the end user it is important to know the format expected in the yaml configuration file. Maybe changing the title to "Marker Configuration File Format", making these points a list and add a small sentence about what to expect might help, e.g. "Observe that this `yaml` configuration file follows a specific format: .... ".

tttthomasssss commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+The top level contains unique user-specified names for each marker. This is a dictionary mapping marker names to marker definitions.
+
+Each marker definition can contain an arbitrary combination of `conditions` and `operators` that map to a valid sub-marker definition.
```
Whats a sub-marker definition?

aeshky replied:
rephrased this 👍

tttthomasssss commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+Each `operator` is a dictionary mapping an operator tag (e.g., `and`) to a list of sub markers, which can be conditions or a nested operator-conditions definition (e.g., see `marker_mood_expressed_and_name_provided`).
+
+:::note
+You cannot yet reuse an existing marker name in the definition of another marker.
```
Love the subtle soft announcement of a future feature 😆

tttthomasssss commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+Each marker definition can contain an arbitrary combination of `conditions` and `operators` that map to a valid sub-marker definition.
+
+Each `condition` maps a condition tag to a value. For example, in `marker_name_provided`, the tag `slot_was_set` is mapped to the slot `name` which exists in the `domain.yaml` file.
+
+Each `operator` is a dictionary mapping an operator tag (e.g., `and`) to a list of sub markers, which can be conditions or a nested operator-conditions definition (e.g., see `marker_mood_expressed_and_name_provided`).
```
This is fairly dense stuff and I had to read it a few times to understand it (even though I know what it is expressing). Would it make sense to have the full example from above here again in order to highlight the different parts?

tttthomasssss commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+## Running the Markers script
+
+Markers are extracted from the dialogues already stored in your [tracker store](https://rasa.com/docs/rasa/next/tracker-stores). Once you've created your Marker definitions in the Makers configuration file, you can extract them by running the follow command:
```
typo in `[...] Makers configuration file, [...]`

hsm207 replied:
typo in `you can extract them by running the follow command`

tttthomasssss commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+positional arguments:
+  output_filename       The filename to write the extracted markers to (CSV format).
+  {first_n,sample,all}
+    first_n             Select trackers sequentially until N are taken.
+    sample              Select trackers by sampling N.
```
Is there a way to specify `N` for `first_n` and `sample`?

aeshky replied:
yes there is a `count` parameter but I don't see that info here. Well spotted.

aeshky replied:
You can get info about `count` and `seed` by typing:
`rasa evaluate markers output.csv sample --help`
However, I've raised an [issue](https://github.com/RasaHQ/rasa/issues/10174) to modify this. It will likely change to:
`rasa evaluate markers <strategy> --help`
where strategy is one of `all`, `sample_n` or `first_n`

tttthomasssss reviewed (changes requested):
This is looking great 🚀  - I have added some comments.

ka-bu commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+We extract information about top-level markers which have user-specified names, and extract meta information about all the events where the markers applied.
+
+For the `at_least_once` marker we only extract meta information about the first occurrence.
```
It might be good to re-arrange the sections a bit and move everything that gives an high-level overview on what can be achieved with the marker stuff to the first part (along with a running example) and leave the implementation bits for later, because details like these (and the implications) might be overlooked easily otherwise.
- So what could be moved to the top:  "relevant events" that we're interested per marker, what meta data we currently look at, and what we analyse. The running example in that section can explain the relevant events without mentioning the actual output format, and we can also explain how to answer some questions with our "analysis". 
- What could be left for later: configuration file format, how to extract / compute statistics, how the output is formatted

ka-bu commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,291 @@
+
+```
+sender_id,session_idx,marker,statistic,value
+all,nan,-,total_number_of_sessions,3
````
Might be good to add a note explaining that the "nan" is not a bug -- and do we explain the "session" aspect anywhere or do we expect everyone to know this?

koaning replied:
I'm not representative of every user out there, but I can confirm that my first response to seeing `nan` displayed as "expected" caused me to feel a bit bamboozled. I'd for sure appreciate add a sentence or two.

koaning commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,291 @@
+Markers are extracted from the dialogues already stored in your [tracker store](https://rasa.com/docs/rasa/next/tracker-stores). Once you've created your Marker definitions in the Makers configuration file, you can extract them by running the follow command:
+
+```bash
+rasa evaluate markers markers.yaml extracted_markers.csv all
````
Are we sure this command is correct? 

```
> rasa evaluate markers markers.yaml extracted_markers.csv all

usage: rasa evaluate markers [-h] [-v] [-vv] [--quiet] [--config CONFIG]
                             [--no-stats | --stats-file-prefix [STATS_FILE_PREFIX]]
                             [--endpoints ENDPOINTS] [-d DOMAIN]
                             output_filename {first_n,sample,all} ...
rasa evaluate markers: error: argument strategy: invalid choice: 'extracted_markers.csv' (choose from 'first_n', 'sample', 'all')
```

Also per `--help`. 

```
rasa evaluate markers --help                                

usage: rasa evaluate markers [-h] [-v] [-vv] [--quiet] [--config CONFIG]
                             [--no-stats | --stats-file-prefix [STATS_FILE_PREFIX]]
                             [--endpoints ENDPOINTS] [-d DOMAIN]
                             output_filename {first_n,sample,all}
```

The main positional argument here should be `extracted_markers.csv`, no? That means we shouldn't add `markers.yaml` in this line.

usc-m replied:
```rasa evaluate markers --config markers.yaml extracted_markers.csv all``` should work (for argparse reasons, and because the strategies are subparsers, optional args on the base command have to be before the strategy)

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+optional arguments:
+  -h, --help            show this help message and exit
+  --config CONFIG       The config file(s) containing marker definitions. This can be a single YAML
```
I'm wondering if `--config` is the best name here since it may cause confusion with `config.yaml`. Might it be possible to rename this to `--markers`?

aeshky replied:
We're considering this under this [issue](https://github.com/RasaHQ/rasa/issues/10174). I think it's fine to have different 'configurations' for differen things. I worry we'll end up using 'markers' to mean both the marker config and the marker output 😄

koaning commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,291 @@
+                             [--endpoints ENDPOINTS] [-d DOMAIN]
+                             output_filename {first_n,sample,all} ...
+
+positional arguments:
+  output_filename       The filename to write the extracted markers to (CSV format).
+  {first_n,sample,all}
+    first_n             Select trackers sequentially until N are taken.
+    sample              Select trackers by sampling N.
+    all                 Select all trackers.
+
+optional arguments:
+  -h, --help            show this help message and exit
+  --config CONFIG       The config file(s) containing marker definitions. This can be a single YAML
+                        file, or a directory that contains several files with marker definitions in
+                        it. The content of these files will be read and merged together. (default:
+                        markers.yml)
+  --no-stats            Do not compute summary statistics. (default: True)
+  --stats-file-prefix [STATS_FILE_PREFIX]
+                        The common file prefix of the files where we write out the compute
+                        statistics. More precisely, the file prefix must consist of a common path
+                        plus a common file prefix, to which suffixes `-overall.csv` and `-per-
+                        session.csv` will be added automatically. (default: stats)
+  --endpoints ENDPOINTS
+                        Configuration file for the tracker store as a yml file. (default:
+                        endpoints.yml)
+  -d DOMAIN, --domain DOMAIN
+                        Domain specification. This can be a single YAML file, or a directory that
+                        contains several files with domain specifications in it. The content of
+                        these files will be read and merged together. (default: domain.yml)
+
+Python Logging Options:
+  -v, --verbose         Be verbose. Sets logging level to INFO. (default: None)
+  -vv, --debug          Print lots of debugging statements. Sets logging level to DEBUG. (default:
+                        None)
+  --quiet               Be quiet! Sets logging level to WARNING. (default: None)
+```
````
Likely a typo: has this been copied twice in the same doc?

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+stats-overall.csv
+
+## Configuring the CLI command
```
Should this be part of the markers doc or part of the [cli docs](https://rasa.com/docs/rasa/command-line-interface/)?

aeshky replied:
I like having them here, but no strong preference from me. Alex proposed to add the full usage instructions here. Do you think we should move them?

aeshky replied:
Moved it to CLI page.

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+all,nan,marker_mood_expressed,number_of_sessions_where_marker_applies_at_least_once,2
+all,nan,marker_mood_expressed,percentage_of_sessions_where_marker_applies_at_least_once,66.667
+all,nan,marker_mood_expressed_and_name_provided,number_of_sessions_where_marker_applies_at_least_once,0
+all,nan,marker_mood_expressed_and_name_provided,percentage_of_sessions_where_marker_applies_at_least_once,0.0
```
Do we really expect a `nan` as a `session_id` in the csv-file? If they're all `nan`, I'm wondering why the column wasn't omitted. Or is this happening only in a specific setting?

ka-bu replied:
there is a nan because trackers is "all" and `session_idx`s are ints so we shouldn't use "all" to indicate that this is over all sessions easily -- we could move those rows to another file to get rid of the column entirely, use "all" and accept that this column will be loaded as "str" in pandas, ... or use a different value in that column ?

aeshky replied:
I think I just need to explain what the `nan`s mean here. I like having the same column names for both statistics output files. 🤔

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+all,nan,marker_cheer_up,percentage_of_sessions_where_marker_applies_at_least_once,0.0
+all,nan,marker_mood_expressed,number_of_sessions_where_marker_applies_at_least_once,2
+all,nan,marker_mood_expressed,percentage_of_sessions_where_marker_applies_at_least_once,66.667
+all,nan,marker_mood_expressed_and_name_provided,number_of_sessions_where_marker_applies_at_least_once,0
+all,nan,marker_mood_expressed_and_name_provided,percentage_of_sessions_where_marker_applies_at_least_once,0.0
```
Is the `value` column an int or a float? Seeing two different schemas here is a bit scary.

aeshky replied:
It will depend on the statistic calculated. For example, "the number of" was written as an `int` while "percentage" was written as a `float`. If you read it as a pandas dataframe they will all be treated as floats. I think it's fine as long as the column only contains numbers. What do you think?

koaning replied:
I worry about the pandas assumption. Yes, pandas will deal with it in an expected way. But what about excel? Spark? Dask? Polars? 

Each library typically has its own little quirk when it comes to casting assumptions and anything that we can proactively do to prevent assumptions from taking over may prevent a small disaster. There's a (pretty darn good) [PyData Amsterdam talk](https://youtu.be/hoaMzebS_l4?t=517) titled "High Performance Data Loss" on the topic.

aeshky replied:
We can create an issue for this and include this in the upcoming marker fixes, as it's more to do with the code than the docs. Thanks for raising it.

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+Thus we compute the following:
+
+1. **For each session and each marker** we compute "per-session statistics" which include the arithmetic mean, median, minimum, and maximum number of user turns preceding the event  at which the marker applied
```
A space too many right after "event".

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+### Extracted Markers
+
+We extract information about top-level markers which have user-specified names, and extract meta information about all the events where the markers applied.
```
This section uses the term "we" frequently and this is a pattern that goes against [the documentation language guidelines](https://www.notion.so/rasa/Tone-and-language-guidelines-cd96f21292b24c1ab7f7457c7973ec69#766266eaf5744ca094bd247409684d28). 

It's about the user, so maybe rephrase it as; 

```
We extract information -> You can extract information
```

hsm207 replied:
> extract information about top-level markers which have user-specified names

Is there such a thing as a top-level marker that does not have a user-specified name?

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+:::
+
+Markers allow you to describe points in a dialogue that you are interested in identifying. In Rasa, a dialogue is represented as a sequence of events, which include bot actions that were executed (`ActionExecuted`), intents that were detected (`UserUttered`), and slots that were set (`SlotSet`). A point of interest in a dialogue can thus be expressed as a logical expression that can be evaluated  ...
```
The start of the paragraph talks about what "you" can do. While the bottom talks about what "we" can do. Since the docs are about the user who reads it, I propose using "you" over "we" for consistency. 

`we "mark" that point in the dialogue` -> `you can "mark" that point in the dialogue`.

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+### Conditions
+
+We support the following conditions:
```
~~We~~ `->` Marker files support the following conditions.

koaning replied:
Come to think of it. Maybe "You can use the following conditions in marker files." is even better.

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+- `intent`: the specified user intent was detected.
+- `slot_was_set`: the specified slot was set.
+
+We also support their negated forms:
```
~~We~~ `->` Negated forms of these conditions are also supported.

koaning replied:
Come to think of it. Maybe "You can use the negated forms of these conditions too." is even better.

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+### Operators
+
+To combine conditions, we provide the following operators:
```
Another `we` that could be omitted. My suggestion may be to change

> To combine conditions, ~~we~~ provide the following operators:

into 

> Combinations can be combined by using operators. The following operators are supported.

aeshky replied:
I'll go over all of these instances and make the document consistent. Thanks.

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+For the `at_least_once` marker we only extract meta information about the first occurrence.
+
+We extract the following meta information:
```
We -> "You can extract the following meta information"

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+1. the number of user turns (i.e., the number of `UserUttered` events) preceding the event at which a marker applied
+2. the index of the event at which a marker applied
+
+We output the extracted information in a `.csv` file that contains the following columns:
```
We -> "You can see the extracted information in a `.csv` file with the following columns:

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+We output the extracted information in a `.csv` file that contains the following columns:
+
+- sender_id
+- session_idx
+- marker name
+- event_idx (where we index all events in the tracker, starting with index 0)
+- num_preceeding_user_turns
```
The `.csv` file below shows all the column names already, so if we prefer to keep this portion I might suggest using it as a legend where each column name is followed by what the user can expect from the values. 

This may also be a great place to explain what `nan` values may be expected.

koaning commented:
I've given the docs the first-order review. There are some comments on style, most of which revolve around the use of the word "we" vs. "you". I've reviewed using the guidelines [on Notion](https://www.notion.so/rasa/Tone-and-language-guidelines-cd96f21292b24c1ab7f7457c7973ec69#766266eaf5744ca094bd247409684d28).

I'm wondering if it makes sense to add a segment to this document that explains how to run this locally. I may be wrong, but I think we can set the `endpoints.yml` to use SQLite locally for this feature. I tried running this and I didn't succeed, so I made an [issue](https://github.com/RasaHQ/rasa/issues/10099) to double-check if this feature is supported. If it is, would it make sense to mention it here? I'll write a tutorial for the docs too, but it'd be a shame if there's a false impression that one couldn't run this unless it is in production.

hsm207 commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,291 @@
+
+To define one or more marker, you need to specify a `yaml` configuration file. Here is an example:
+
+```yaml
````
why don't we also have an example for `never` and `not`?

aeshky commented:
Thanks for your review @koaning 

>There are some comments on style, most of which revolve around the use of the word "we" vs. "you". I've reviewed using the guidelines on Notion.

Thanks for the link. Will follow the guideline. 

> I'm wondering if it makes sense to add a segment to this document that explains how to run this locally. I may be wrong, but I think we can set the endpoints.yml to use SQLite locally for this feature. I tried running this and I didn't succeed, so I made an issue to double-check if this feature is supported. If it is, would it make sense to mention it here? I'll write a tutorial for the docs too, but it'd be a shame if there's a false impression that one couldn't run this unless it is in production. 

It should run locally and I can clarify this in the docs. I created [these steps](https://docs.google.com/document/d/1KkFARn_4cbGOhiQG7D50w5RluiVqYw6A76Y2JzcHBF8/edit) for QA testing using SQLite (perhaps you want to give it a try?).

koaning commented:
> It should run locally and I can clarify this in the docs. I created these steps for QA testing using SQLite (perhaps you want to give it a try?).

@aeshky are you working from RC1 or the main branch? If it's in main I'll certainly check it, but I may prefer to wait until RC02 is out for stability.

aeshky commented:
> are you working from RC1 or the main branch? If it's in main I'll certainly check it, but I may prefer to wait until RC02 is out for stability.

@koaning the instructions I linked are for RC1. 
RC2 is planned for Monday and I'll need to update the QA tests / instructions with the new tags :) 

By the way, will start addressing the comments properly tomorrow now that everyone has had a chance to review.

kedz commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,245 @@
+And here is a marker for surfacing dialogues where all mandatory slots are set except `travel_flight_class`:
+
+```yaml
+marker_dialogue_completion:
````
Perhaps this should have a different name, e.g. `marker_dialogue_incomplete`?

aeshky commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+id: markers
+sidebar_label: Markers
+title: Markers
+description: Read more about how to mark points of interest in dialogues using logical expressions.
```
I'll rephrase it slightly differently 👍

aeshky commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,291 @@
+
+There are several applications for Markers. For example, they can be used to define your bot's Key Performance Indicators (KPIs), such as dialogue completion or task success. For [Carbon Bot](https://rasa.com/blog/using-conversation-tags-to-measure-carbon-bots-success-rate/) (which helps users offset their carbon emissions from flying) dialogue completion can be defined as "all mandatory slots ha ...
+
+Markers also allow you to diagnose your dialogues by surfacing important events for further inspection. For example, we might observe that Carbon Bot tends to successfully set the `travel_departure` and `travel_destination` slots, but not the `travel_flight_class`. We can define a marker to quantify how often this scenario occurs and surface relevant dialogues for review as part of [Conversation  ...
```
We don't currently have a plan to do that as part of the release. I know research are working with Carbon using older prototypes of markers and they might end up switching. No concrete plan though.

However, we do plan to add a marker config to the `rasa init` project to help people get started. The examples below also use the `rasa init` bot so people can copy and modify them. 

I chose carbon in the into because it illustrates business KPIs more clearly than mood bot.

aeshky replied:
If you would like to have a branch on Carbon Bot with markers we can create an issue for it and add it to the research board.

hsm207 replied:
Thanks for the explanation. I think having `rasa init` add the marker config is good enough.

aeshky commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,302 @@
+Here is a sample of the extracted markers output file:
+
+```
+sender_id,session_idx,marker_name,event_id,num_preceding_user_turns
````
@usc-m and @ka-bu I noticed that we use marker_name in the extracted marker output, but use marker in the stats files. I would like to make it consistent if that's okay, switching to marker for both.

Not sure if I should do it in this PR or in another one (possibly combining it with any CLI changes). What do you think?

usc-m replied:
Sounds good - I'd split it out into it's own PR to not clutter this one too much (or combine with the CLI changes)

aeshky commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,302 @@
+---
+id: markers
+sidebar_label: Markers
+title: Markers
```
@hsm207 where should we put this page? Under `Concepts`?

aeshky replied:
I placed it under concepts :)

tttthomasssss commented on docs/docs/command-line-interface.mdx:
```diff
@@ -23,6 +23,7 @@ abstract: The command line interface (CLI) gives you easy-to-remember commands f
 |`rasa data convert`     |Converts training data between different formats.                                                                                         |
 |`rasa data validate`    |Checks the domain, NLU and conversation data for inconsistencies.                                                                                             |
 |`rasa export`           |Exports conversations from a tracker store to an event broker.                                                                            |
+|`rasa evaluate markers` |Applies markers to existing tracker store.                                                                                                |
```
> Applies markers to existing tracker store.

I think this is a bit too vague (and I'm slightly confused that the command is `rasa evaluate markers`, but the explanation says `Applies`, from the command I'd expect that it gives me some sort of performance measure?)

aeshky replied:
There was a discussion [here](https://rasa-hq.slack.com/archives/C01HHMR4X8S/p1634044314193800) about what the call the command. To summarise, we expect to release a number of evaluation tools, the first of which is Markers.

I'll rephrase this 👍

tttthomasssss commented on docs/docs/command-line-interface.mdx:
```diff
@@ -419,6 +420,53 @@ This command is most commonly used to import old conversations into Rasa X to an
+
+:::
+
+This command allows you to apply the [markers](./markers.mdx) you defined in your marker configuration file,
+to pre-existing dialogues stored in your [tracker store](./tracker-stores.mdx):
```
I think it would be helpful if this also explains something about the expected output of the operation.

tttthomasssss commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,289 @@
+which include bot actions that were executed, intents that were detected, and slots that were set.
+Markers allow you to describe conditions over such events, and when these conditions are met, the dialogues are "marked" for further analysis or inspection.
+
+There are several applications for Markers. For example, they can be used to define your bot's **Key Performance Indicators (KPIs)**,
```
Is uppercase Markers a typo?

aeshky replied:
I use upper case when I'm referring to the feature, and lowercase when I'm referring to individual markers. Does this make sense or should I lowercase all of them?

tttthomasssss replied:
Right, I have no strong opinions either way 🙂

tttthomasssss commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,289 @@
+      - slot_was_set: travel_flight_class
+```
+
+The next sections explain how write marker definitions, how to apply them to your existing dialogues, and what the output format looks like.
````
missing `to` (?) in "[...] explain how write marker definitions, [...]"

tttthomasssss commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,289 @@
+
+## Defining Markers
+
+You define you markers in a marker configuration file written in `YAML`.
```
typo, in `You define you markers [...]` - should presumably be `your markers`.

tttthomasssss commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,289 @@
+```
+The unique marker identifier is `marker_mood_expressed`. This marker definition contains one operator `or`,
+and two event conditions `intent: mood_unhappy` and `intent: mood_great`.
+
````
Maybe add a sentence explaining what this means, i.e. that the `steedmarker` is triggered when _either_ `mood unhappy` is detected _or_ `mood great` is detected.

tttthomasssss commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,289 @@
+
+## Extracting Markers
+
+Markers are extracted from dialogues already stored a tracker store. To learn how to store interactions with your bot in a tracker store,
```
missing `in` in "[...] from dialogues already stored [in] a tracker store."

tttthomasssss reviewed (commented):
This looks awesome 🤖 , I only found a bunch of very minor things.

aeshky commented on docs/docs/marker.mdx:
```diff
@@ -0,0 +1,280 @@
+
+Markers allow you to describe points in a dialogue that you are interested in identifying. In Rasa, a dialogue is represented as a sequence of events, which include bot actions that were executed (`ActionExecuted`), intents that were detected (`UserUttered`), and slots that were set (`SlotSet`). A point of interest in a dialogue can thus be expressed as a logical expression that can be evaluated  ...
+
+There are several applications for Markers. For example, they can be used to define your bot's Key Performance Indicators (KPIs), such as dialogue completion or task success. For [Carbon Bot](https://rasa.com/blog/using-conversation-tags-to-measure-carbon-bots-success-rate/) (which helps users offset their carbon emissions from flying) dialogue completion can be defined as "all mandatory slots ha ...
```
I tried to give a flavour of this with the paragraph you're highlighting. I can go into more depth in the blog post. What do you think?

koaning commented:
When I read the docs today I noticed that I understood what "number of preceding user turns" means. But it occurred to me why this metric is interesting. How would a conversational designer use this statistic to improve their bot? 

After toying around I think it's related to "if it takes a long time, then the user may be misunderstood a few times". If this is the case, it deserves to be mentioned explicitly as a use case in the documentation.

aeshky commented:
>When I read the docs today I noticed that I understood what "number of preceding user turns" means. But immediately after it occurred to me that I didn't see why this metric is interesting. How would a conversational designer use this statistic to improve their bot?

The number of preceding user turns is telling you how many utterances the user had to make before reaching the event that you care about. It's a good measure of progress in task-oriented dialogues from the perspective of the user. For example if you had a marker that defines task success you want to know how many turns it took to reach success. Like you say, maybe the user had to rephrase multiple times which caused the dialogue to become longer, or maybe the task wasn't clear so the user had to ask clarifying questions. The number of turns is used as a metric dialogue research (here are some examples [[1](https://scholar.google.com/scholar?cluster=15047530170291380940&hl=en&as_sdt=0,5)], [[2](https://www.sciencedirect.com/science/article/pii/S0167639308000459)], [[3](https://www.tandfonline.com/doi/abs/10.1080/08839514.2013.835230)]). If the dialogues are 1:1 (1 system utterance for every 1 user utterance) you'll sometimes see the "number of turns" measured, but I think the "number of user turns" is more meaningful anyway, and especially when the system can make multiple utterances / take multiple actions (as is the case with Rasa). 

The pure number of events is less meaningful because you can have a lot of system events that are not part of the dialogue (querying a database, starting a session, etc.). So the number of user turns give a better intuition of how long the dialogue is and particularly from the user's perspective. 

Does this make sense? I almost added an explanation to the document but felt that it was moving away from usage instruction and wondered if it's better in a blog post or a tutorial. But since you raised it, I think it's worth giving the intuition in the docs too.

aeshky commented:
@koaning I added an explanation here: https://github.com/RasaHQ/rasa/commit/f2275ccad1853db202eb9d35b8a4dec875840d0c
Let me know what you think.

koaning commented:
Yep! Those two paragraphs are much appreciated 😃 !

koaning commented on docs/docs/command-line-interface.mdx:
```diff
@@ -419,6 +420,54 @@ This command is most commonly used to import old conversations into Rasa X to an
+
+:::caution
+
+This feature is currently experimental and might change or be removed in the future. Share your feedback in the forum to help us make it production-ready.
```
Do we have a thread on the forum? It cannot hurt to start a thread right now and point people to it. This way much of the feedback comes in one place.

aeshky replied:
I haven't started a thread, and don't know if someone else has (maybe @TyDunn?). 
Do we have guidelines about this in notion?

TyDunn replied:
I have not. Is this something you can work on with @EmmaWightman, @aeshky? I'd happy to review the content before it goes live as well as support you through the process as much as you feel you need :)

aeshky replied:
Sounds good! I'll contact Emma.

koaning replied:
You're free to contact Emma, but for the Tensorflow 2.6 upgrade, I just took the liberty of starting a thread [on the forum myself](https://forum.rasa.com/t/feedback-upgrading-to-tensorflow-2-6/48331). I can also help you with that if you'd like (Emma's going to be very busy soon with the advent of all the social announcements).

aeshky replied:
Thanks both. How about I start a thread by reusing the content posted [here under "Experimental Feature: Markers"](https://rasa.com/blog/what-s-ahead-in-rasa-open-source-3-0/?utm_campaign=Rasa%20Open%20Source%203.0%20launch&utm_content=187975830&utm_medium=social&utm_source=twitter&hss_channel=tw-2839825685)?

aeshky replied:
[Here](https://docs.google.com/document/d/1OcLvhSsVpGFTSBcHd1YBJFcoZms7lRr5XpzHiSu2o3Q/edit?usp=sharing) is an abridged version.

If you're happy with it, I'll post it on the forum and tag myself and maybe vincent?

koaning replied:
Text looks good to me 👍  Feel free to start a thread, but I wouldn't share it on Twitter just yet. My impression is that at the moment we're mainly getting feedback from our superhero community because they're guaranteed to be familiar with Rasa. We can share via community slack tho!

TyDunn replied:
@aeshky Can we wait to share until after the 3.0 release / live event about markers?

aeshky replied:
@TyDunn Sorry, I already posted it [here](https://forum.rasa.com/t/feedback-markers-a-new-experimental-feature-in-rasa/49344). 😅
I think it might be good if people read a bit about the feature beforehand. We might get better questions that. What do you think? I can remove the post if you disagree :)

TyDunn replied:
I still don't think we should post it yet. Can we take it down until the release is out and we have a chance to tell people about it? The timing will be better if we post it right before we have the live event for it

aeshky replied:
Ok done!

aeshky replied:
@TyDunn are you happy for me to post this at the end of the day, prior to the release event on Monday?

TyDunn replied:
@aeshky For sure! Now that 3.0 is released it makes sense 🚀

koaning commented on docs/docs/command-line-interface.mdx:
```diff
@@ -419,6 +420,54 @@ This command is most commonly used to import old conversations into Rasa X to an
+:::
+
+The following command applies the [markers](./markers.mdx) you defined in your marker configuration file,
+to pre-existing dialogues stored in your [tracker store](./tracker-stores.mdx), and produces `CSV` files containing
```
```suggestion
to pre-existing dialogues stored in your [tracker store](./tracker-stores.mdx), and produces `.csv` files containing

```

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,321 @@
+
+:::caution
+
+This feature is currently experimental and might change or be removed in the future. Share your feedback in the forum to help us make it production-ready.
```
Same as mentioned earlier, it's better to have a forum link.

koaning commented on docs/docs/markers.mdx:
```diff
@@ -0,0 +1,321 @@
+`marker_mood_expressed`, the intents `mood_unhappy` and `mood_unhappy` are both intents listed in the mood bot's `domain.yml` file.
+
+:::note
+You cannot yet reuse an existing marker name in the definition of another marker.
```
```suggestion
You cannot reuse an existing marker name in the definition of another marker.
```

koaning replied:
Having "yet" in there implies that we will have it in the future, best not to announce anything in an experimental feature.

koaning commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,321 @@
+you can apply your markers to your trackers by running the following command:
+
+```bash
+rasa evaluate markers all --config marker.yml extracted_markers.csv
````
The CLI section suggests that we assume `markers.yml` as a default value. Would be best to stick to that filename.

koaning replied:
```suggestion
rasa evaluate markers all --config markers.yml extracted_markers.csv

```

koaning commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,321 @@
+```bash
+rasa evaluate markers all --config marker.yml extracted_markers.csv
+```
+This script will process the marker definitions you provide your in the marker configuration file: `marker.yml`.
````
```suggestion
This script will process the marker definitions you provide in the marker configuration file: `markers.yml`.
```

koaning commented on docs/docs/markers.mdx:
````diff
@@ -0,0 +1,321 @@
+4d55093e9696452c8d1157fa33fd54b2,0,marker_cheer_up_failed,14,2
+c00b3de97713427d85524c4374125db1,0,marker_mood_expressed,2,0
+```
+Each row represents an occurrence of the marker specified under the `marker` column.
````
Don't we list this per `sender_id` and `session_idx`?

```suggestion
Each row represents an occurrence of the marker specified under the `marker` column, for each `sender_id` and `session_idx`. 
```
